### PR TITLE
Display error message if the primary email is not verified

### PR DESCRIPTION
**Pull request by dojutsu-user**
The work is in progress.
#4025

### CODE REVIEW

**Comment by dojutsu-user** [comment]
@humitos The work is in progress. 
I have addded the error message, if the primary email of the user is not verified.

screenshot from 2018-10-16 01-56-45

I will fix the tests.
Also, the added CSS is not in correct place, I will take care of it.

**Comment by humitos** [comment]
The user will need a way to take action: click a button/click a link to the admin page/etc where the email is re-sent to him. Otherwise, the user can't verify the email address.

**Comment by dojutsu-user** [comment]
@humitos hey, I need some advice.
I was looking for the reasons for failing tests, and they were failing because of this.

```python3
def get_context_data(self, **kwargs):
        context = super(ProjectDashboard, self).get_context_data(**kwargs)

        primary_email = self.request.user.emailaddress_set.filter(primary=True).first()        
        context['primary_email'] = primary_email.email
        context['primary_email_verified'] = primary_email.verified

        return context
```

The tests are using `django_dynamic_fixture`
```python3
import django_dynamic_fixture as fixture
user = fixture.get(User)
```
so the dummy user generated by this have no email set as primary email, so these lines when executed
```python3
self.client.force_login(user)
response = self.client.get('/dashboard/')
```
fails the tests with the error `None type as no attribute email` which is traced back to the method `get_context_data`.
So to prevent tests from failing, an `if` condition will work like this
```python3
def get_context_data(self, **kwargs):
        context = super(ProjectDashboard, self).get_context_data(**kwargs)

        primary_email = self.request.user.emailaddress_set.filter(primary=True).first()
        if primary_email:
            context['primary_email'] = primary_email.email
            context['primary_email_verified'] = primary_email.verified

        return context
```

So will it be okay to do so??

**Comment by stsewd** [comment]
If you need an email for the tests, you can create that in the fixture

```python
import django_dynamic_fixture as fixture
user = fixture.get(User, email='....')
```

**Comment by dojutsu-user** [comment]
@humitos I tried it earlier (from the docs of django_dynamic_fixture), but the tests still fail with the same error.

**Comment by dojutsu-user** [comment]
@stsewd 
@humitos 

I was finding the reasons for the above failing tests.
Apparently, this line 
```python3
import django_dynamic_fixture as fixture
user = fixture.get(User, email='xyz@xyz.com')
```
is not registering the email of the user as primary or unverified. Probably, this is using `create` method to create the instances.
```python3
User.objects.create(username="xyz", email="xyz@xyz.com")
```
and I found that it does not register the email in the model `allauth.account.models.EmailAddress`.
This reason is not clear to me.

But, I found a way to get it work.
```python3
import django_dynamic_fixture as fixture
from allauth.account.models import EmailAddress

user = fixture.get(User)
EmailAddress.objects.create(user=user, email='this_is_temporary_email@temp.com', primary=True)
```
Is it okay?

**Comment by stsewd** [comment]
I haven't looked at the code for user creation, but I think it's ok. I'll try to check later

**Comment by stsewd** [comment]
I haven't looked at the code for user creation, but I think it's ok. I'll try to check later

**Comment by dojutsu-user** [comment]
@stsewd Would you mind helping me a bit. I am not able to correct the tests failing in `rtd_tests/tests/test_privacy_urls.py` and `rtd_tests/tests/test_project_views.py`.
Although, I was able to fix other tests with the help of the method described in [above comment](https://github.com/rtfd/readthedocs.org/pull/4758#issuecomment-430420346)

File: readthedocs/projects/views/private.py [hunk 1]
```diff
@@ -61,6 +61,10 @@ def get_queryset(self):
     def get_context_data(self, **kwargs):
         context = super(ProjectDashboard, self).get_context_data(**kwargs)
 
+        primary_email = self.request.user.emailaddress_set.filter(primary=True).first()
+        context['primary_email'] = primary_email.email
+        context['primary_email_verified'] = primary_email.verified
```
[the remark this answers is not in the document]
Comment (dojutsu-user) [reply readthedocs/projects/views/private.py]: > pass the whole object here

That means, pass the `context` object as parameter to the `get_context_data` method and then use it inside the function.
Is this what you're saying?

[the remark this answers is not in the document]
Comment (stsewd) [reply readthedocs/projects/views/private.py]:
```
context['primary_email'] = primary_email
```

[the remark this answers is not in the document]
Comment (dojutsu-user) [reply readthedocs/projects/views/private.py]: yeahhh... rest is clear to me.
Thanks very much.

**Comment by dojutsu-user** [comment]
@stsewd Can you please give your review to the UI shown in the comment [above](https://github.com/rtfd/readthedocs.org/pull/4758#issuecomment-430307496)

**Comment by dojutsu-user** [comment]
@stsewd Can you please review this.

**Comment by stsewd** [comment]
@ericholscher do we have users without a primary email? The Anonymous user doesn't have one (local dev), I guess in the past that wasn't a requirement

**Comment by ericholscher** [comment]
Believe everyone should have a primary email, though I'm not 100% sure.

**Comment by dojutsu-user** [comment]
@ericholscher 
@stsewd 
I think that each user has a primary email (in the production). The Signup form has Email field as a required field. And for the user signing up with Github/GitLab/BitBucket get their primary email from these accounts only.
Also when try to remove primary email, it shows me this error
screenshot from 2018-11-02 10-11-29

**Comment by dojutsu-user** [comment]
Final Result

screenshot from 2018-11-02 10-11-29

File: readthedocs/templates/projects/project_dashboard.html [hunk 2]
```diff
@@ -1,5 +1,27 @@
...
+        <form action="{% url 'account_email' %}" method="post">
+          {% csrf_token %}
+          <input type="hidden" value="{{ primary_email.email }}" name="email" />
+          <button type="submit" name="action_send">Re-send activation link</button>
+        </form>
+      </div>
+    {% endif %}
+  {% else %}
+      <div class="verification-error-block">
+        Please Add A Primary Email. Click <a href="{% url 'account_email' %}">Here</a> To Add.
```
Comment (stsewd) [inline readthedocs/templates/projects/project_dashboard.html]: We should use sentence case as we do with the another msg

**Comment by dojutsu-user** [comment]
@agjohnson 
I have made another PR which uses the inbuilt Notification system for showing this error.
#4964

**Comment by humitos** [comment]
Closing this PR in favor of #4964 which uses our own notification system.

**Pull request closed without merging**